test/test_distributions.py

```diff
@@ -616,6 +624,59 @@ def test_chi2_sample_grad(self):
                                        'rel error {}'.format(rel_error),
                                        'max error {}'.format(rel_error.max())]))
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_studentT_shape(self):
+        df = Variable(torch.exp(torch.randn(2, 3)), requires_grad=True)
+        df_1d = Variable(torch.exp(torch.randn(1)), requires_grad=True)
+        self.assertEqual(StudentT(df).sample().size(), (2, 3))
+        self.assertEqual(StudentT(df).sample_n(5).size(), (5, 2, 3))
+        self.assertEqual(StudentT(df_1d).sample_n(1).size(), (1, 1))
+        self.assertEqual(StudentT(df_1d).sample().size(), (1,))
+        self.assertEqual(StudentT(0.5).sample().size(), (1,))
+        self.assertEqual(StudentT(0.5).sample_n(1).size(), (1,))
+
+        def ref_log_prob(idx, x, log_prob):
+            d = df.data.view(-1)[idx]
+            expected = scipy.stats.t.logpdf(x, d)
+            self.assertAlmostEqual(log_prob, expected, places=3)
+
+        self._check_log_prob(StudentT(df), ref_log_prob)
+
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_studentT_sample(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
+        for df in [0.1, 1.0, 5.0]:
+            self._check_sampler_sampler(StudentT(df),
+                                        scipy.stats.t(df),
+                                        'StudentT(df={})'.format(df))
+
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_studentT_sample_grad(self):
```

Comment:
I believe this test cannot pass because the reparameterized gradients are stochastic and are only correct in expectation. It is safe to remove these tests.

The reason the gradients are stochastic is that our sampler contains a nuisance variable (projects the pair `X`, `Z` down to a single number). To compute zero-variance reparameterized gradients we would need to explicitly compute the cdf gradient. That's a lot of work (see implementations of `standard_gamma_grad()` and `dirichlet_grad()`). Actually my original implementation of `Beta` and `Dirichlet` had stochastic gradients for the same reason: I simply relied on `Gamma.rsample()` but that contains the nuisance variable `gammas.sum(-1)` which is projected out during normalization. I really wanted zero-variance gradients for `Dirichlet` (so we could implement e.g. LDA) so I put a bunch of work into implementing `dirichlet_grad()`.

Reply:
To see this, you should be able to plot reparameterized gradients as a function of df, using quantities computed in this test:
```py
dfs = Variable(10 ** torch.arange(-2, 4, 0.01))
actual_grad = ...
expected_grad = ...
pyplot.plot(df, expected_grad, 'k--', label='truth')
pyplot.scatter(df, actual_grad, label='approx')
pyplot.xscale('log')
pyplot.yscale('log')
pyplot.legend()
```

Reply:
Also see example plots (click on Detals) in the Dirichlet PR description: https://github.com/probtorch/pytorch/pull/28